src/Stratis.Bitcoin.Features.Consensus/CoinViews/CachedCoinView.cs

```diff
@@ -393,37 +366,162 @@ public async Task<uint256> Rewind()
+                this.persistedBlockHash = hash;
                 this.blockHash = hash;
 
                 this.logger.LogTrace("(-):'{0}'", hash);
                 return hash;
             }
         }
 
+        /// <summary>
+        /// Dequeues the rewind data continuously and saves it to the database when max batch size is reached or timer ran out.
+        /// </summary>
+        /// <remarks>Batch is always saved on shutdown.</remarks>
+        private async Task DequeueRewindDataContinuouslyAsync()
+        {
+            this.logger.LogTrace("()");
+
+            Task<QueuedRewindData> dequeueTask = null;
+            Task timerTask = null;
+
+            while (!this.nodeLifetime.ApplicationStopping.IsCancellationRequested)
+            {
+                // Start new dequeue task if not started already.
+                dequeueTask = dequeueTask ?? this.rewindDataQueue.DequeueAsync();
+
+                // Wait for one of the tasks: dequeue or timer (if available) to finish.
+                Task task = (timerTask == null) ? dequeueTask : await Task.WhenAny(dequeueTask, timerTask).ConfigureAwait(false);
+
+                bool saveBatch = false;
+
+                try
+                {
+                    await task.ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    // Happens when node is shutting down or Dispose() is called.
+                    // We want to save whatever is in the batch before exiting the loop.
+                    saveBatch = true;
+
+                    this.logger.LogDebug("Node is shutting down. Save batch.");
+                }
+
+                // Save batch if timer ran out or we've dequeued a new rewind data 
+                // or the max batch size is reached or the node is shutting down.
+                if (dequeueTask.Status == TaskStatus.RanToCompletion)
+                {
+                    QueuedRewindData item = dequeueTask.Result;
+
+                    // Set the dequeue task to null so it can be assigned on the next iteration.
+                    dequeueTask = null;
+
+                    this.rewindDataBatch.Add(item);
+
+                    this.currentCountOfAllRewindDataItems += item.ItemsCount;
+
+                    saveBatch = saveBatch || (this.currentCountOfAllRewindDataItems >= BatchMaxNumberOfRewindDataItems) || this.chainState.IsAtBestChainTip;
+                }
+                else
+                {
+                    // Will be executed in case timer ran out or node is being shut down.
+                    saveBatch = true;
+                }
+
+                if (saveBatch)
+                {
+                    if (this.rewindDataBatch.Count != 0)
+                    {
+                        await this.SaveBatchAsync().ConfigureAwait(false);
+
+                        this.rewindDataBatch.Clear();
```

Comment:
There is an assumption the `this.SaveBatchAsync()` has worked.  If it fails or returns before saving the code is always clearing `this.rewindDataBatch`.

Is there any mileage is getting some sort of return value from `this.SaveBatchAsync()` and if all is well clear `this.rewindDataBatch`, and set `this.currentCountOfAllRewindDataItems = 0`. ?